core/src/main/scala/kafka/tools/StreamsResetter.java

```diff
@@ -300,23 +350,160 @@ private void maybeSeekToEnd(final KafkaConsumer<byte[], byte[]> client,
+                final Long timestamp = now.getTime();
+
+                final Map<TopicPartition, Long> topicPartitionsAndTimes = new HashMap<>(inputTopicPartitions.size());
+                for (final TopicPartition topicPartition : inputTopicPartitions) {
+                    topicPartitionsAndTimes.put(topicPartition, timestamp);
+                }
+
+                final Map<TopicPartition, OffsetAndTimestamp> topicPartitionsAndOffset = client.offsetsForTimes(topicPartitionsAndTimes);
+
+                for (final TopicPartition topicPartition : inputTopicPartitions) {
+                    final Long offset = topicPartitionsAndOffset.get(topicPartition).offset();
+                    client.seek(topicPartition, offset);
+                }
+            } else if (options.has(fromFileOption)) {
+                final Map<TopicPartition, Long> endOffsets = client.endOffsets(inputTopicPartitions);
+                final Map<TopicPartition, Long> beginningOffsets = client.beginningOffsets(inputTopicPartitions);
+
+                final String resetPlanPath = options.valueOf(fromFileOption);
+                final String resetPlanCsv = Utils.readFileAsString(resetPlanPath);
+                final Map<TopicPartition, Long> topicPartitionsAndOffset = parseResetPlan(resetPlanCsv);
+
+                final Map<TopicPartition, Long> validatedTopicPartitionsAndOffset =
+                    checkOffsetRange(topicPartitionsAndOffset, beginningOffsets, endOffsets);
+
+                for (final TopicPartition topicPartition : inputTopicPartitions) {
+                    final Long offset = validatedTopicPartitionsAndOffset.get(topicPartition);
+                    client.seek(topicPartition, offset);
+                }
+            } else {
+                client.seekToBeginning(inputTopicPartitions);
+            }
+
+            for (final TopicPartition p : inputTopicPartitions) {
+                final Long position = client.position(p);
+                System.out.println("Topic: " + p.topic() + " Partition: " + p.partition() + " Offset: " + position);
+            }
+        }
+    }
+
+    public Long getDateTime(String ts) throws ParseException {
+        if (!(ts.split("T")[1].contains("+") || ts.split("T")[1].contains("-") || ts.split("T")[1].contains("Z"))) {
+            ts = ts + "Z";
+        }
+
+        Date date;
+        try {
+            date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").parse(ts);
+        } catch (ParseException e) {
+            date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX").parse(ts);
+        }
+
+        return date.getTime();
+    }
+
+    private Map<TopicPartition, Long> parseResetPlan(String resetPlanCsv) {
+        final Map<TopicPartition, Long> topicPartitionAndOffset = new HashMap<>();
+        for (final String line : resetPlanCsv.split("\n")) {
+            final String[] parts = line.split(",");
+            final String topic = parts[0];
```

Comment:
Should we guard against NPE? (same blow)

Reply:
Yeap, I'll refactor.